trees/avltree/avltree.go

```diff
@@ -0,0 +1,493 @@
+			n = n.c[0]
+		case c > 0:
+			n = n.c[1]
+		}
+	}
+	if found {
+		return
+	}
+	return nil, false
+}
+
+// Put inserts node into the tree.
+// Key should adhere to the comparator's type assertion, otherwise method panics.
+func (t *Tree) Put(key interface{}, value interface{}) {
+	var put func(*Node, **Node) bool
+	put = func(p *Node, qp **Node) bool {
+		q := *qp
+		if q == nil {
+			t.size++
+			*qp = &Node{Key: key, Value: value, p: p}
+			return true
+		}
+
+		c := t.comparator(key, q.Key)
+		if c == 0 {
+			q.Key = key
+			q.Value = value
+			return false
+		}
+
+		if c < 0 {
+			c = -1
+		} else {
+			c = 1
+		}
+		a := (c + 1) / 2
+		var fix bool
+		fix = put(q, &q.c[a])
+		if fix {
+			return putFix(int8(c), qp)
+		}
+		return false
+	}
+
+	put(nil, &t.Root)
+}
+
+// Remove remove the node from the tree by key.
+// Key should adhere to the comparator's type assertion, otherwise method panics.
+func (t *Tree) Remove(key interface{}) {
+	var remove func(**Node) bool
+	remove = func(qp **Node) bool {
+		q := *qp
+		if q == nil {
+			return false
+		}
+
+		c := t.comparator(key, q.Key)
+		if c == 0 {
+			t.size--
+			if q.c[1] == nil {
+				if q.c[0] != nil {
+					q.c[0].p = q.p
+				}
+				*qp = q.c[0]
+				return true
+			}
+			fix := removeMin(&q.c[1], &q.Key, &q.Value)
+			if fix {
+				return removeFix(-1, qp)
+			}
+			return false
+		}
+
+		if c < 0 {
+			c = -1
+		} else {
+			c = 1
+		}
+		a := (c + 1) / 2
+		fix := remove(&q.c[a])
+		if fix {
+			return removeFix(int8(-c), qp)
+		}
+		return false
+	}
+
+	remove(&t.Root)
+}
+
+func removeMin(qp **Node, minKey *interface{}, minVal *interface{}) bool {
+	q := *qp
+	if q.c[0] == nil {
+		*minKey = q.Key
+		*minVal = q.Value
+		if q.c[1] != nil {
+			q.c[1].p = q.p
+		}
+		*qp = q.c[1]
+		return true
+	}
+	fix := removeMin(&q.c[0], minKey, minVal)
+	if fix {
+		return removeFix(1, qp)
+	}
+	return false
+}
+
+func putFix(c int8, t **Node) bool {
+	s := *t
+	if s.b == 0 {
+		s.b = c
+		return true
+	}
+
+	if s.b == -c {
+		s.b = 0
+		return false
+	}
+
+	if s.c[(c+1)/2].b == c {
+		s = singlerot(c, s)
+	} else {
+		s = doublerot(c, s)
+	}
+	*t = s
+	return false
+}
+
+func removeFix(c int8, t **Node) bool {
+	s := *t
+	if s.b == 0 {
+		s.b = c
+		return false
+	}
+
+	if s.b == -c {
+		s.b = 0
+		return true
+	}
+
+	a := (c + 1) / 2
+	if s.c[a].b == 0 {
+		s = rotate(c, s)
+		s.b = -c
+		*t = s
+		return false
+	}
+
+	if s.c[a].b == c {
+		s = singlerot(c, s)
+	} else {
+		s = doublerot(c, s)
+	}
+	*t = s
+	return true
+}
+
+func singlerot(c int8, s *Node) *Node {
+	dbgLog.Printf("singlerot: enter %p:%v %d\n", s, s, c)
+	s.b = 0
+	s = rotate(c, s)
+	s.b = 0
+	dbgLog.Printf("singlerot: exit %p:%v\n", s, s)
+	return s
+}
+
+func doublerot(c int8, s *Node) *Node {
+	dbgLog.Printf("doublerot: enter %p:%v %d\n", s, s, c)
+	a := (c + 1) / 2
+	r := s.c[a]
+	s.c[a] = rotate(-c, s.c[a])
+	p := rotate(c, s)
+	if r.p != p || s.p != p {
+		panic("doublerot: bad parents")
```

Comment:
You are not missing anything. It will never happen. Feel free to remove it but I always shy to leaving those things in.